### PR TITLE
feat: make Geant4 surface mapper configurable

### PR DESCRIPTION
**Pull request by asalzburger**
This PR breaks the direct dependency of `ActsExamples::Geant4Simulation` from `TrackingGeometry` and allows to set a customised SensitiveSurface mapper.

The current SensitiveSurface mapper would fail for e.g. ITK geometry because the center comparison would probably not be valid. Also, it allows to test with Gen2 geometry, which will be needed (for the moment) for the `traccc` input simulation.

### CODE REVIEW

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp [hunk 1]
```diff
@@ -9,6 +9,8 @@
 #pragma once
 
 #include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/Utilities/GridAccessHelpers.hpp"
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp:13]: unused?

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp [hunk 2]
```diff
@@ -62,10 +67,12 @@ class SensitiveSurfaceMapper {
   /// hierarchy and applies name remapping to the Physical volumes
   /// of the Geant4 geometry.
   ///
+  /// @param gctx the geometry context
   /// @param g4PhysicalVolume the current physical volume in process
   /// @param motherPosition the absolute position of the mother
   /// @param sCounter  a counter of how many volumes have been remapped
-  void remapSensitiveNames(G4VPhysicalVolume* g4PhysicalVolume,
+  void remapSensitiveNames(const Acts::GeometryContext& gctx,
```
Comment (andiwand) [inline Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp:74]: why do we need the geometry context here? does that make sense to pass it around during simulation?

File: Examples/Python/src/Geant4Component.cpp [hunk 3]
```diff
@@ -97,6 +98,88 @@ PYBIND11_MODULE(ActsPythonBindingsGeant4, mod) {
...
+    auto sm =
+        py::class_<SensitiveSurfaceMapper,
+                   std::shared_ptr<SensitiveSurfaceMapper>>(
+            mod, "SensitiveSurfaceMapper")
+            .def(py::init([](const Config& cfg, Acts::Logging::Level level) {
+              return std::make_shared<SensitiveSurfaceMapper>(
+                  cfg, getDefaultLogger("SensitiveSurfaceMapper", level));
+            }));
+
+    auto c = py::class_<SensitiveSurfaceMapper::Config>(sm, "Config")
```
Comment (andiwand) [inline Examples/Python/src/Geant4Component.cpp:112]:
```suggestion
    auto c = py::class_<Config>(sm, "Config")
```

File: Examples/Python/src/Geant4Component.cpp [hunk 4]
```diff
@@ -97,6 +98,88 @@ PYBIND11_MODULE(ActsPythonBindingsGeant4, mod) {
...
+                 .def(py::init<>());
+    ACTS_PYTHON_STRUCT_BEGIN(c, SensitiveSurfaceMapper::Config);
```
Comment (andiwand) [inline Examples/Python/src/Geant4Component.cpp:114]:
```suggestion
    ACTS_PYTHON_STRUCT_BEGIN(c, Config);
```

File: Examples/Python/src/Geant4Component.cpp [hunk 5]
```diff
@@ -97,6 +98,88 @@ PYBIND11_MODULE(ActsPythonBindingsGeant4, mod) {
...
+    ACTS_PYTHON_MEMBER(volumeMappings);
+    ACTS_PYTHON_MEMBER(candidateSurfaces);
+    ACTS_PYTHON_STRUCT_END();
+
+    sm.def(
+        "create", [](const Config& cfg, Acts::Logging::Level level,
+                     const std::shared_ptr<const TrackingGeometry> tGeometry) {
+          // Helper struct to find the sensitive surface candidates
+          struct SensitiveCandidates {
+            std::shared_ptr<const TrackingGeometry> trackingGeometry;
```
Comment (andiwand) [inline Examples/Python/src/Geant4Component.cpp:125]: Would it make sense to have helpers for that in C++?

Having the code here forces you to rewrite it if you are calling G4 from C++

File: Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp (hunk 1 above)
Comment (asalzburger) [reply Examples/Algorithms/Geant4/include/ActsExamples/Geant4/SensitiveSurfaceMapper.hpp:13]: And removed.

File: Examples/Python/src/Geant4Component.cpp (hunk 3 above)
Comment (asalzburger) [reply Examples/Python/src/Geant4Component.cpp:112]: We ask in the matching for center() which takes a GeometryContext, the Matcher should get one, actually correctly also the setup of the geometry should use one.

File: Examples/Python/src/Geant4Component.cpp (hunk 4 above)
Comment (asalzburger) [reply Examples/Python/src/Geant4Component.cpp:114]: Fixed.

File: Examples/Python/src/Geant4Component.cpp (hunk 5 above)
Comment (asalzburger) [reply Examples/Python/src/Geant4Component.cpp:125]: I pulled the TrackingGeometry part back into C++, not the `Experimental::Detector` one. We want to test the experimental detector w/o dragging the `Experimental` code everywhere for the moment.

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [992d86de6152ac79fdd79715ceb4af75d2b45ace]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522413)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522412)||
|:green_circle: | [test_ActsPersistifySeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522411)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522410)||
|:green_circle: | [test_ActsAnalogueClustering](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522409)||
|:green_circle: | [test_ActsConversionWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522408)||
|:green_circle: | [test_ActsWorkflowHeavyIons](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522407)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522406)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522405)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522404)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522403)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522402)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522401)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522400)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522399)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522398)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522397)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522396)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522395)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522394)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522393)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522392)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522391)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522390)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522389)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522388)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522387)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522386)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/38522385)||

**Pull request merged**